courses/dss/labrpc/src/network.rs

```diff
@@ -0,0 +1,355 @@
+        } = end_info;
+
+        match (enabled, server) {
+            (true, Some(server)) => {
+                let short_delay = if !reliable {
+                    // short delay
+                    let ms = thread_rng().gen::<u64>() % 27;
+                    Some(ms)
+                } else {
+                    None
+                };
+
+                if !reliable && (thread_rng().gen::<u64>() % 1000) < 100 {
+                    // drop the request, return as if timeout
+                    Delay::new(Duration::from_secs(short_delay.unwrap())).await;
+                    return Err(Error::Timeout);
+                }
+
+                let drop_reply = !reliable && thread_rng().gen::<u64>() % 1000 < 100;
+                let long_reordering = if long_reordering && thread_rng().gen_range(0, 900) < 600i32
+                {
+                    // delay the response for a while
+                    let upper_bound: u64 = 1 + thread_rng().gen_range(0, 2000);
+                    Some(200 + thread_rng().gen_range(0, upper_bound))
+                } else {
+                    None
+                };
+
+                // Dispatch
+                process_rpc(
+                    short_delay,
+                    drop_reply,
+                    long_reordering,
+                    rpc,
+                    network,
+                    server,
+                )
+                .await
+            }
+            _ => {
+                // simulate no reply and eventual timeout.
+                let ms = if self.core.long_delays.load(Ordering::Acquire) {
+                    // let Raft tests check that leader doesn't send
+                    // RPCs synchronously.
+                    thread_rng().gen::<u64>() % 7000
+                } else {
+                    // many kv tests require the client to try each
+                    // server in fairly rapid succession.
+                    thread_rng().gen::<u64>() % 100
+                };
+
+                debug!("{:?} delay {}ms then timeout", rpc, ms);
+                Delay::new(Duration::from_millis(ms)).await;
+                Err(Error::Timeout)
+            }
+        }
+    }
+
+    /// Spawns a future to run on this net framework.
+    pub fn spawn<F>(&self, f: F)
+    where
+        F: Future<Output = ()> + Send + 'static,
+    {
+        self.core.worker.spawn_ok(f);
+    }
+
+    /// Spawns a future to run on this net framework.
+    pub fn spawn_poller<F>(&self, f: F)
+    where
+        F: Future<Output = ()> + Send + 'static,
+    {
+        self.core.poller.spawn_ok(f);
+    }
+}
+
+async fn process_rpc(
+    delay: Option<u64>,
+    drop_reply: bool,
+    long_reordering: Option<u64>,
+    mut rpc: Rpc,
+    network: Network,
+    server: Server,
+) -> Result<Vec<u8>> {
+    // Dispatch ===============================================================
+    if let Some(delay) = delay {
+        Delay::new(Duration::from_secs(delay)).await;
+    }
```

Comment:
Looks like we need to take the delay to prevent polling it twice?

https://github.com/pingcap/talent-plan/blob/2a1f4c20241c4b58f21a91ca83eb4b331a924377/courses/dss/labrpc/src/lib.rs#L598-L600